torch/fx/passes/reinplace.py

```diff
@@ -0,0 +1,303 @@
+                attr_itr = getattr(attr_itr, atom)
+            return attr_itr
+
+        for node in self.graph.nodes:
+            if node.op == 'placeholder':
+                result = next(args_iter)
+            elif node.op == 'get_attr':
+                result = fetch_attr(node.target)
+            elif node.op == 'call_function':
+                result = node.target(*load_arg(node.args), **load_arg(node.kwargs))
+            elif node.op == 'call_method':
+                self_obj, *args = load_arg(node.args)
+                kwargs = load_arg(node.kwargs)
+                result = getattr(self_obj, node.target)(*args, **kwargs)
+            elif node.op == 'call_module':
+                result = self.modules[node.target](*load_arg(node.args), **load_arg(node.kwargs))
+
+            # Maintain mapping from each tensor in the program to its name in the graph
+            tensor_to_node_name_map[result] = node.name
+
+            for a in itertools.chain(node.args, node.kwargs.values()):
+                arg = load_arg(a)
+                if isinstance(arg, torch.Tensor):
+                    for i, inpt in enumerate(args):
+                        if isinstance(inpt, torch.Tensor) and is_view_of(arg, inpt):
+                            input_aliases[i].add(a.name)
+
+                    # Keep track of, for every "base" tensor in the program, what are all of its aliases.
+                    curr_base_tensor_name = a.name if arg._base is None else tensor_to_node_name_map[arg._base]
+                    base_tensor_to_aliases_map[curr_base_tensor_name].add(a.name)
+
+            env[node.name] = result
+
+        # Post-process our base-tensor-to-aliases info into a more convenient lookup table
+        alias_map = defaultdict(set)
+        for base_tensor_name, alias_names in base_tensor_to_aliases_map.items():
+            for name in [base_tensor_name] + list(alias_names):
+                alias_map[name] = alias_names
+
+        return input_aliases, alias_map
+
+
+def schemas_match(functional_schema, inplace_schema):
+    names_match = inplace_schema.name.endswith("_") and inplace_schema.name[:-1] == functional_schema.name
+    arg_types_match = all(
+        a1.type == a2.type for a1, a2 in zip(functional_schema.arguments, inplace_schema.arguments))
+    # for the inplace op, its first argument should be mutable
+    assert inplace_schema.arguments[0].alias_info is not None and inplace_schema.arguments[0].alias_info.is_write
+    # and its remaining arguments shouldn't be.
+    assert all(a.alias_info is None for a in inplace_schema.arguments[1:])
+    return names_match and arg_types_match
+
+@compatibility(is_backward_compatible=True)
+def reinplace(gm: fx.GraphModule, *sample_args, **sample_kwargs):
+    """
+    Given an fx.GraphModule, modifies it to perform "reinplacing".
+    We look for out-of-place op call sites like `b = a.add(...)`,
```

Comment:
What's assumption of IR for this pass? 
`b = a.add(...)` seems to imply that it's working on torch IR, but it should also work for aten IR... 

or any functional IR?